jax/_src/scipy/special.py

```diff
@@ -724,82 +725,249 @@ def i1(x: ArrayLike) -> Array:
+def j1_small(x: ArrayLike) -> Array:
+    '''
+    Implementation of J1 for x < 5 
+    '''
+    z = x * x
+    w = jnp.polyval(jnp.array(RP1), z) / jnp.polyval(jnp.array(RQ1), z)
+    w = w * x * (z - jnp.array(Z1)) * (z - jnp.array(Z2))
+    return w
+
+def j1_large(x: ArrayLike) -> Array:  
+    '''
+    Implementation of J1 for x > 5 
+    '''  
+    w = 5.0 / x
+    z = w * w
+    p = jnp.polyval(jnp.array(PP1), z) / jnp.polyval(jnp.array(PQ1), z)
+    q = jnp.polyval(jnp.array(QP1), z) / jnp.polyval(jnp.array(QQ1), z)
+    xn = x - THPIO4
+    p = p * jnp.cos(xn) - w * q * jnp.sin(xn)
+    return p * SQ2OPI / jnp.sqrt(x)
+
+def bessel_j1(z: ArrayLike) -> Array:
+    """
+    Bessel function of the first kind of order one and a real argument 
+    - using the implementation from CEPHES, translated to Jax, to match scipy to machine precision.
+
+    Reference:
+    Cephes mathematical library.
+
+    Args:
+        x: The sampling point(s) at which the Bessel function of the first kind are
+        computed.
+
+    Returns:
+        An array of shape `x.shape` containing the values of the Bessel function
+    """
+
+    z = jnp.asarray(z)
+    z, = promote_dtypes_inexact(z)
+    z_dtype = lax.dtype(z)
+
+    if dtypes.issubdtype(z_dtype, complex):
+      raise ValueError("complex input not supported.")
+
+    return jnp.sign(z)*jnp.where(jnp.abs(z) < 5.0, j1_small(jnp.abs(z)),j1_large(jnp.abs(z)))
+
+def j0_small(x: ArrayLike) -> Array:
+    '''
+    Implementation of J0 for x < 5 
+    '''
+    z = x * x
+    # if x < 1.0e-5:
+    #     return 1.0 - z/4.0
+
+    p = (z - jnp.array(DR10)) * (z - jnp.array(DR20))
+    p = p * jnp.polyval(jnp.array(RP0),z)/jnp.polyval(jnp.array(RQ0), z)
+    return jnp.where(x<1e-5,1-z/4.0,p)
+    
+
+def j0_large(x: ArrayLike) -> Array:
+    '''
+    Implementation of J0 for x >= 5    
+    '''
+
+    w = 5.0/x
+    q = 25.0/(x*x)
+    p = jnp.polyval(jnp.array(PP0), q)/jnp.polyval(jnp.array(PQ0), q)
+    q = jnp.polyval(jnp.array(QP0), q)/jnp.polyval(jnp.array(QQ0), q)
+    xn = x - PIO4
+    p = p * jnp.cos(xn) - w * q * jnp.sin(xn)
+    return p * SQ2OPI / jnp.sqrt(x)
+
+def bessel_j0(z: ArrayLike) -> Array:
+    """
+    Bessel function of the first kind of order zero and a real argument 
+    - using the implementation from CEPHES, translated to Jax, to match scipy to machine precision.
+
+    Reference:
+    Cephes Mathematical Library.
+
+    Args:
+        z: The sampling point(s) at which the Bessel function of the first kind are
+        computed.
+
+    Returns:
+        An array of shape `x.shape` containing the values of the Bessel function
+    """
+    z = jnp.asarray(z)
+    z, = promote_dtypes_inexact(z)
+    z_dtype = lax.dtype(z)
+
+    if dtypes.issubdtype(z_dtype, complex):
+      raise ValueError("complex input not supported.")
+
+    return jnp.where(jnp.abs(z) < 5.0, j0_small(jnp.abs(z)),j0_large(jnp.abs(z)))
+    
+def bessel_jn(z: ArrayLike, v: int) -> Array:
```

Comment:
It would be nice here if we could match the API of `scipy.special.jn`. It's a bit different in that the order of arguments is switched, and the `v` argument can have an arbitrary shape. Implementing it for a dynamic scalar `v` and wrapping the function in `jax.numpy.vectorize` may be the easiest approach, though algorithmically it could be useful to specially handle the case where `v` is a static scalar int.

Reply:
I think I might need help with anything clever here about the vectorization - I don't think I know how to do this nicely. As for the real valued v, we need the modified Bessel Iv functions to do this. Perhaps we drop this right now in this pull request and loop back to it later? It's pretty trivial to manually construct the integer orders by scan, so maybe it doesn't need to be in core jax.

Reply:
I don't think a real-valued `v` is necessary, but supporting a non-scalar integer would be useful. You should be able to use `jax.numpy.vectorize` to do this if you change your `scan` to `fori_loop`.

Reply:
Do we lose some of the advantages of `scan` that way?

Reply:
Neat, didn't know that. I actually worry we might have to drop higher order Jn: recurrence relations are numerically unstable for high order [ref](https://www.sciencedirect.com/science/article/abs/pii/0021999176900322). I don't have time right now but the right approach seems to be like you suggest - implement the full scipy real-z `jv` function.

Suggest dropping this right now and merging the PR with only `j0`, `j1` and can revisit this later? Actually only need 0 and 1 for most optics problems.